src/connector/src/parser/postgres.rs

```diff
@@ -93,284 +92,252 @@ pub fn postgres_row_to_owned_row(row: tokio_postgres::Row, schema: &Schema) -> O
+                // instead of `Option<EnumString>` to handle enum to keep the behaviors aligned.
+                // An enum list contains `NULL` will be converted to `NULL`.
+                let res = row.try_get::<_, Option<Vec<EnumString>>>(i);
+                match res {
+                    Ok(val) => {
+                        if let Some(v) = val {
+                            v.into_iter().for_each(|val| {
+                                builder.append(Some(ScalarImpl::from(val.0)));
+                            });
                         }
+                        Some(ScalarImpl::from(ListValue::new(builder.finish())))
+                    }
+                    Err(err) => {
+                        log_error!(name, err, "parse enum column failed");
+                        None
                     }
                 }
-                DataType::Jsonb => {
-                    handle_data_type!(row, i, name, serde_json::Value, JsonbVal)
-                }
-                DataType::Interval => {
-                    handle_data_type!(row, i, name, Interval)
-                }
-                DataType::List(dtype) => {
-                    let mut builder = dtype.create_array_builder(0);
-                    // enum list needs to be handled separately
-                    if let Kind::Array(item_type) = row.columns()[i].type_().kind()
-                        && let Kind::Enum(_) = item_type.kind()
-                    {
-                        // FIXME(Kexiang): The null of enum list is not supported in Debezium.
-                        // As `NULL` in enum list is not supported in Debezium, we use `EnumString`
-                        // instead of `Option<EnumString>` to handle enum to keep the behaviors aligned.
-                        // An enum list contains `NULL` will be converted to `NULL`.
-                        let res = row.try_get::<_, Option<Vec<EnumString>>>(i);
+            } else {
+                match **dtype {
+                    DataType::Boolean => {
+                        handle_list_data_type!(row, i, name, bool, builder);
+                    }
+                    DataType::Int16 => {
+                        handle_list_data_type!(row, i, name, i16, builder);
+                    }
+                    DataType::Int32 => {
+                        handle_list_data_type!(row, i, name, i32, builder);
+                    }
+                    DataType::Int64 => {
+                        handle_list_data_type!(row, i, name, i64, builder);
+                    }
+                    DataType::Float32 => {
+                        handle_list_data_type!(row, i, name, f32, builder);
+                    }
+                    DataType::Float64 => {
+                        handle_list_data_type!(row, i, name, f64, builder);
+                    }
+                    DataType::Decimal => {
+                        let res = row.try_get::<_, Option<Vec<Option<PgNumeric>>>>(i);
```

Comment:
An idea to simplify the code:
We can add a `List(Vec<Option<ScalarAdapter<'a>>>)` variant to the `ScalarAdapter` in
https://github.com/risingwavelabs/risingwave/pull/16589/files#diff-affc26b39891b665d7ecfdfab3a8a0e57390db9d370eeca2e54ab6a60a881458R78-R83

The this part can be refactored to 
```rust
let res = row.try_get::<_, Option<ScalarAdapter>>(i);
match res {
     Ok(val) => {
         val.map(|adapter| adapter.into_scalar(DataType::Decimal) )
     }
}
```

So that we can encapsulate these convert logic into `ScalarAdapter::into_scalar`.

Reply:
Worth a try. To clarify my https://github.com/risingwavelabs/risingwave/pull/16589#discussion_r1596279363 is only to document the status of that PR, rather than suggesting `List(Vec<Option<ScalarAdapter<'a>>>)` is a bad idea.